# Pre-Processing

This notebook contains the necessary pre-processing steps.

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
pd.set_option('display.max_columns', 32)

In [2]:
#Loading data
bookings = pd.read_csv('post_eda_bookings.csv')
bookings.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0      0            0          7               2015               July   
1      0            0         13               2015               July   
2      0            0         14               2015               July   
3      0            0          0               2015               July   
4      0            0          9               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     1       1       0.0       0   
1                        0                     1       1       0.0       0   
2                        0                     2       2       0.0       0   
3                        0                     2       2       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment  is_repeated_guest  previous_cancellations  \
0   BB     GBR         Direct                  0                       0   
1   BB     GBR      Corporate                  0                       0   
2   BB     GBR      Online TA                  0                       0   
3   BB     PRT         Direct                  0                       0   
4   FB     PRT         Direct                  0                       0   

   previous_bookings_not_canceled reserved_room_type assigned_room_type  \
0                               0                  A                  C   
1                               0                  A                  A   
2                               0                  A                  A   
3                               0                  C                  C   
4                               0                  C                  C   

   booking_changes deposit_type  agent  days_in_waiting_list customer_type  \
0                0   No Deposit   -1.0                     0     Transient   
1                0   No Deposit  304.0                     0     Transient   
2                0   No Deposit  240.0                     0     Transient   
3                0   No Deposit   -1.0                     0     Transient   
4                0   No Deposit  303.0                     0     Transient   

     adr  required_car_parking_spaces  total_of_special_requests  \
0   75.0                            0                          0   
1   75.0                            0                          0   
2   98.0                            0                          1   
3  107.0                            0                          0   
4  103.0                            0                          1   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-02  
1          Check-Out              2015-07-02  
2          Check-Out              2015-07-03  
3          Check-Out              2015-07-03  
4          Check-Out              2015-07-03

## Feature Reduction

In [3]:
#Exploring columns
bookings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 86249 entries, 0 to 86248
Data columns (total 30 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           86249 non-null  int64  
 1   is_canceled                     86249 non-null  int64  
 2   lead_time                       86249 non-null  int64  
 3   arrival_date_year               86249 non-null  int64  
 4   arrival_date_month              86249 non-null  object 
 5   arrival_date_week_number        86249 non-null  int64  
 6   arrival_date_day_of_month       86249 non-null  int64  
 7   stays_in_weekend_nights         86249 non-null  int64  
 8   stays_in_week_nights            86249 non-null  int64  
 9   adults                          86249 non-null  int64  
 10  children                        86249 non-null  float64
 11  babies                          86249 non-null  int64  
 12  meal                            

In [4]:
#Looking at the number of unique values for categorical features. Including agent id column
cat = list(bookings.select_dtypes(include=['object']).columns)
cat.append('agent')
bookings[cat].nunique()

arrival_date_month          12
meal                         4
country                    178
market_segment               8
reserved_room_type          10
assigned_room_type          12
deposit_type                 3
customer_type                4
reservation_status           3
reservation_status_date    924
agent                      334
dtype: int64

I'll take a look at the country column and try to reduce the number of unique values. I'm also taking a look at the agent column. The agent column is a numeric id for the travel agency that booked the reservation. I want to see if there's a particular agency that's used more frequently than others. I'm getting rid of reservation_statatus_date and the reservation_status columns. The relevant information for what happened with a reservation is already stored in the is_canceled column. 

**I'll also get rid of the agent column. The agent column is a numeric id for the travel agency that booked the reservation. The type of travel agency is likely to be more useful, and we have that information in the market_segment column.**

In [5]:
#Dropping unnecessary
df= bookings.drop(['reservation_status', 'reservation_status_date'], axis=1)

In [6]:
#Looking at the count of the values
df['agent'].value_counts().head()

 9.0      28655
 240.0    12899
-1.0      11854
 14.0      3330
 7.0       3280
Name: agent, dtype: int64

In [7]:
#Looking at the percentages
(df['agent'].value_counts()/len(df)).head()

 9.0      0.332236
 240.0    0.149555
-1.0      0.137439
 14.0     0.038609
 7.0      0.038029
Name: agent, dtype: float64

In [11]:
df['market_segment'][df['agent']==9]

209         Direct
4993        Direct
11020    Corporate
11369    Corporate
11370    Corporate
           ...    
86238    Online TA
86239    Online TA
86245    Online TA
86246    Online TA
86248    Online TA
Name: market_segment, Length: 28655, dtype: object

It looks like the majority of the reservations were made through the agencies with id 9 and id 240. The -1 represent missing values. Wondering if it would be a good idea to make the column binary. 1 representing all the reservations made through agencies 9 and 240 and 0 for the rest.

In [6]:
#Exploring unique values of the country column
df['country'].value_counts().head(10)

PRT    26618
GBR    10378
FRA     8788
ESP     7192
DEU     5369
ITA     3049
IRL     3011
BEL     2074
BRA     1987
NLD     1907
Name: country, dtype: int64

In [7]:
#Looking at the percentages of the values
(df['country'].value_counts()/len(df)).head(10)

PRT    0.308618
GBR    0.120326
FRA    0.101891
ESP    0.083386
DEU    0.062250
ITA    0.035351
IRL    0.034911
BEL    0.024047
BRA    0.023038
NLD    0.022110
Name: country, dtype: float64

In [9]:
#Grouping the countries that account for less than 1% of the column into 'Other'
pct = (df['country'].value_counts()/len(df))
mask = df['country'].isin(pct[pct<0.01].index)
df['country'][mask] = 'Other'
df['country'].nunique()

## Dummy encoding

In [10]:
#Defining X
X = df.drop('is_canceled', axis=1)

#Dummy encoding X
X = pd.get_dummies(X, drop_first=True)

In [11]:
#Looking at how many columns I'll be working with
X.shape

(86249, 78)

In [12]:
#Defining my response variable. Using ravel() to get a flattened array
y= df['is_canceled'].ravel()

## Splitting the data into training set and test set

In [13]:
#Leaving 25% of my data for testing
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.25, random_state=12)

## Scaling the data

In [14]:
#Instantiating a scaler
sc = StandardScaler()

#Fitting the scaler to the training set and transforming it
X_train_scaled = sc.fit_transform(X_train)

#Scaling test set
X_test_scaled = sc.transform(X_test)